### PR TITLE
Plugins v2

### PR DESCRIPTION
**Pull request by regalialong**
Our new approach is that we encapsulate all the AssetsTools objects in a `ContentFileContainer` object, which is then passed to Plugins and Blocklist patch.

I took a profile comparing https://github.com/AdGoBye/AdGoBye/commit/35893791d6067b8ec4debe5ff1ce9c679e65dbed (main) against https://github.com/AdGoBye/AdGoBye/commit/a2a8fd64c286841f09e88823d0073d5cce90ced2 (pluginsv2) with ExamplePlugin and our upstream Blocklists at https://github.com/AdGoBye/AdGoBye-Blocklists/commit/5646b6d5aecf00d336184bd70fd4c090b3a25f86. 
PatchContent takes 1,294ms on 35893791d6067b8ec4debe5ff1ce9c679e65dbed while a2a8fd64c286841f09e88823d0073d5cce90ced2 takes 788ms. That's a 40% difference, and this should scale to be even larger depending on the amount of Plugins used.

The diff of ExamplePlugin also shows the non-performance benefit quite nicely, we cut back on the redundant read-write ceremony that Plugins otherwise have to go through.

For parallel patching, I took a profile of 35893791d6067b8ec4debe5ff1ce9c679e65dbed (main) vs 6075d61bae57245c3a6be56790461074015be6d8 (pluginsv2).
I needed to wrap the actual patching loop into a void Patch() function in main so that I can profile the execution time of that, I grabbed 18 worlds from AGBBase and ran it with both ExamplePlugin and upstream blocklists similarly as above.

35893791d6067b8ec4debe5ff1ce9c679e65dbed takes 15,316ms while 6075d61bae57245c3a6be56790461074015be6d8 takes 3,647ms which is a 76% difference. This was embarrassingly parallelizable. Same scaling notice also applies here.

### CODE REVIEW

**Comment by regalialong** [comment]
Okay, I think we need usability improvements since we do a lot of things in parallel but dump them into the console indiscriminately but that should be for another branch. I think this should be workable to put into main unless someone wants another function for v2.

File: AdGoBye.Plugins/IPlugin.cs [hunk 1]
```diff
@@ -28,13 +27,39 @@ public interface IPlugin
...
+    EVerifyResult Verify(ref readonly ContentFileContainer fileContainer);
+
+    /// <summary>
+    /// Initialize is a function that Plugins can use to set up their environment before executing
+    /// <see cref="Patch"/> and <see cref="Verify"/> functions.
+    /// </summary>
+    void Initialize();
+
+    /// <summary>
+    /// Initialize is a function that Plugins can use to clean up after themselves after execution.
```
Comment (DubyaDude) [inline AdGoBye.Plugins/IPlugin.cs:51]: This is not 'Initialize'

File: AdGoBye.Plugins/IPlugin.cs [hunk 2]
```diff
@@ -28,13 +27,39 @@ public interface IPlugin
...
+    ///     Verify is a non-edit stage where Plugins can run environment and validity checks on the asset before
+    ///     operating on it in <see cref="Patch"/>.
+    /// </summary>
+    /// <param name="fileContainer">Container for the underlying asset being operated on</param>
+    /// <returns>A <see cref="EVerifyResult" /> that signifies the result of the plugin's verify operation.
+    ///          Non-<see cref="EVerifyResult.Success"/> returns will skip this Plugin from being executed. </returns>
+    EVerifyResult Verify(ref readonly ContentFileContainer fileContainer);
+
+    /// <summary>
+    /// Initialize is a function that Plugins can use to set up their environment before executing
```
Comment (DubyaDude) [inline AdGoBye.Plugins/IPlugin.cs:45]: Documentation here isn't clear on when this actually occurs.

Also personally not a big fan of 'what should be done in the event' if it doesn't involve the functionality of the app itself. Might be best to separate that under a remark.

File: AdGoBye.Plugins/IPlugin.cs (hunk 1 above)
Comment (DubyaDude) [inline AdGoBye.Plugins/IPlugin.cs:51]: Documentation here isn't clear on when this actually occurs.

Also personally not a big fan of 'what should be done in the event' if it doesn't involve the functionality of the app itself. Might be best to separate that under a remark.
Is this after each patch, or after everything queued is done? Just specify that.

File: AdGoBye.Plugins/IPlugin.cs [hunk 3]
```diff
@@ -28,13 +27,39 @@ public interface IPlugin
...
+    /// <summary>
+    /// WantsIndexerTracking allows a Plugin to pick if it wants the Indexer to skip it when the Indexer thinks
+    /// the Plugin has already patched the file.
+    /// </summary>
+    /// <remarks>
+    /// When a Plugin patches a file, the Indexer keeps track of the Plugin that modified that version of the file,
+    /// which non-deterministic and exotically designed Plugins may not benifit from.
+    ///</remarks>
+    /// <returns>Boolean indicating if the Indexer should skip if thought already patched.</returns>
+    bool WantsIndexerTracking();
```
Comment (DubyaDude) [inline AdGoBye.Plugins/IPlugin.cs:64]: Personally recommend any 'Configuration related field' should come before events

File: AdGoBye/Indexer.cs [hunk 4]
```diff
@@ -328,25 +339,35 @@ public static void PatchContent(Content content)
...
 
+        container.Bundle.file.BlockAndDirInfo.DirectoryInfos[1].SetNewData(container.AssetsFile.file);
+        using var writer = new AssetsFileWriter(file + ".clean");
+        container.Bundle.file.Write(writer);
+        // Moving the file without closing our access fails on NT.
+        writer.Close();
+        container.Bundle.file.Close();
+        container.Bundle.file.Close();
+
+        // TODO: Provide option to disable backup file?
```
Comment (DubyaDude) [inline AdGoBye/Indexer.cs:367]: TODO present, something you want to finish here or save for later?

File: AdGoBye/PluginLoader.cs [hunk 5]
```diff
@@ -76,16 +73,51 @@ private static void LoadPlugin(string pluginPath)
...
             return;
         }
 
+#if !DEBUG
+        if (!Settings.Options.DisablePluginInstallWarning)
+        {
+            var allowlist = LoadPluginAllowlist();
+            if (!allowlist.Contains($"{pluginName} ({pluginMaintainer}, {pluginVersion})"))
+            {
+                Logger.Information("""
```
Comment (DubyaDude) [inline AdGoBye/PluginLoader.cs:92]: Maybe add a separator instead of spaces?
Maybe like `----...` or `===...`

**Review by DubyaDude — CHANGES_REQUESTED** [review]

File: AdGoBye.Plugins/IPlugin.cs (hunk 1 above)
Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:51]: Oops.

File: AdGoBye/Indexer.cs (hunk 4 above)
Comment (regalialong) [reply AdGoBye/Indexer.cs:367]: I'll do this another time probably but I'm considering adding config fields for disabling the backup file and readding compression.

File: AdGoBye.Plugins/IPlugin.cs (hunk 2 above)
Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:45]: This is kind of a chicken and egg contractual problem, I don't actually know concrete use-cases because feedback has been drip-fed to me, this is essentially only included as meta organization for Plugins at request of #52.

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:45]: Either way, let me rephrase that summary.

File: AdGoBye/PluginLoader.cs (hunk 5 above)
Comment (regalialong) [reply AdGoBye/PluginLoader.cs:92]: Are you suggesting literally something like this for the output?
```suggestion
                Logger.Information("""==================
```

File: AdGoBye.Plugins/IPlugin.cs (hunk 2 above)
Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:45]: Is something like this better?
```suggestion
    /// Initialize is an optional function ran before <see cref="Verify"/> which Plugins may use to prepare their state
    /// before patching a world. It's part of the patching loop and therefore may be called multiple times.
    /// </summary>
```

File: AdGoBye.Plugins/IPlugin.cs (hunk 1 above)
Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs]: Ditto of Initialize, ditto for the rephrasing I could think of:
```c#
/// PostPatch is an optional function ran after <see cref="Patch"/> in the patch loop which Plugins may use to clean
/// their state after patching a world. It's part of the patching loop and therefore may be called multiple times.
```

File: AdGoBye/PluginLoader.cs (hunk 5 above)
Comment (DubyaDude) [reply AdGoBye/PluginLoader.cs:92]: Yup, at the end as well

File: AdGoBye.Plugins/IPlugin.cs (hunk 1 above)
Comment (DubyaDude) [reply AdGoBye.Plugins/IPlugin.cs]: Yea that looks good

**Comment by regalialong** [comment]
Once again, thanks for the help, the fresh eyes help a lot.

File: AdGoBye.Plugins/IPlugin.cs [hunk 6]
```diff
@@ -18,23 +18,49 @@ public interface IPlugin
...
+    /// Initialize is an optional function ran before <see cref="Verify"/> which Plugins may use to prepare their state
+    /// before patching a world. It's part of the patching loop and therefore may be called multiple times.
+    /// </summary>
+    void Initialize();
+
+    /// <summary>
+    /// PostPatch is an optional function ran after <see cref="Patch"/> in the patch loop which Plugins may use to clean
+    /// their state after patching a world. It's part of the patching loop and therefore may be called multiple times.
+    /// </summary>
+    void PostPatch();
```
Comment (DubyaDude) [inline AdGoBye.Plugins/IPlugin.cs:65]: I brought this one up as a possible area for a notification-styled plugin, looking it over it might be nice to get exactly which patch finished the loop. Otherwise, the plugin doesn't know exactly which patch just completed.

**Review by DubyaDude — CHANGES_REQUESTED** [review]

File: AdGoBye.Plugins/IPlugin.cs (hunk 6 above)
Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: We could pass a copy of `Content` for Initialize and PostPatch as that's our exact metadata about the file we'll be operating, we access it as part of checking if the Content is a world.

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: No worries, I've not been working too much on AGB stuff either.

That's more than possible, I just want to avoid passing the content paths to Plugins because I feel like might make them do parsing which we may have done already for them (which is why we give `ContentFileContainer` instead).

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: @DubyaDude Hiya, small bump on this since I figured it went under your radar, this branch is living longer than I'd like.

Comment (DubyaDude) [reply AdGoBye.Plugins/IPlugin.cs:65]: Oh sorry, I assumed you were going forward with passing down the path.
However you decide to do it is fine by me, but it should need passing of some identifier to each of the plugin methods

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: Really ugly, here's what I have staged:
```suggestion
    /// <param name="contentId">ID parsed by the Indexer</param>
    /// <param name="path">Location of the content</param>
    /// <param name="version">Version of the content by </param>
    /// <param name="stableContentName">Hash of the version directory of the content, should be unique</param>
    void PostPatch(string contentId, string path, int version, string stableContentName);
```

Comment (DubyaDude) [reply AdGoBye.Plugins/IPlugin.cs:65]: Hmmm, dont think we'd need the path if were passing in contentid, version, and stable content name. if you'd rather not pass that down there im okay with removing it.

Just be sure this info is given across all method calls (Initialize, Patch, Verify, PostPatch), might be better as its own object class tbh.

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: Yeah, the correctest way to do this would be to pass the `Content` type itself over but it's in the Indexer, so it would need to be extracted into its own package so that both AGB and Plugins can depend on it independently from each other. Then that could just be passed as its own type directly by the Indexer as a sort of context object.

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: Let me experiment with this locally and see how painful this gets.

Comment (regalialong) [reply AdGoBye.Plugins/IPlugin.cs:65]: I was being too whiny, this was fine. Does https://github.com/AdGoBye/AdGoBye/blob/b96977774d14eadf4abe938b8b1fa6f216299681/AdGoBye.Plugins/IPlugin.cs look like something you'd imagine?

Comment (DubyaDude) [reply AdGoBye.Plugins/IPlugin.cs:65]: Yea, that looks good to me
